source/io/module_eggShell.js

```diff
@@ -239,6 +246,257 @@
+            return typeVisitor.visitAnalogWaveform.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitTimestamp = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitTimestamp, 'visitTimestamp');
+            return typeVisitor.visitTimestamp.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitPath = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitPath, 'visitPath');
+            return typeVisitor.visitPath.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitArray = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitArray, 'visitArray');
+            return typeVisitor.visitArray.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitCluster = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitCluster, 'visitCluster');
+            return typeVisitor.visitCluster.apply(typeVisitor, args);
+        };
+
+        var typeDispatchFunctions = [
+            dispatchVisitBoolean,
+            dispatchVisitEnum,
+            dispatchVisitInteger,
+            dispatchVisitFloat,
+            dispatchVisitString,
+            dispatchVisitComplex,
+            dispatchVisitAnalogWaveform,
+            dispatchVisitTimestamp,
+            dispatchVisitPath,
+            dispatchVisitArray,
+            dispatchVisitCluster
+        ];
+
+        var registeredTypes = [
+            Module.eggShell.typeIsBoolean,
+            Module.eggShell.typeIsEnum,
+            Module.eggShell.typeIsInteger,
+            Module.eggShell.typeIsFloat,
+            Module.eggShell.typeIsString,
+            Module.eggShell.typeIsComplex,
+            Module.eggShell.typeIsAnalogWaveform,
+            Module.eggShell.typeIsTimestamp,
+            Module.eggShell.typeIsPath,
+            Module.eggShell.typeIsArray,
+            Module.eggShell.typeIsCluster
+        ];
+
+        Module.eggShell.reflectOnValueRef = publicAPI.eggShell.reflectOnValueRef = function (typeVisitor, valueRef) {
+            if (typeof valueRef !== 'object' || Array.isArray(valueRef)) {
+                throw new Error('valueRef must be an object. Found: ' + valueRef);
+            }
+
+            if (typeof typeVisitor !== 'object' || Array.isArray(typeVisitor)) {
+                throw new Error('typeVisitor must be an object. Found: ' + typeVisitor);
+            }
+
+            var typeRef = valueRef.typeRef,
+                args = Array.prototype.slice.call(arguments, 1),
+                foundTypeIndex = -1,
+                i = 0;
+
+            for (i = 0; i < registeredTypes.length; i += 1) {
+                if (registeredTypes[i](typeRef) === true) {
+                    foundTypeIndex = i;
+                    break;
+                }
+            }
+
+            if (foundTypeIndex < 0) {
+                throw new Error('Unexpected type. Is typeRef pointing to a valid type?. TypeRef found: ' + typeRef);
```

Comment:
Maybe pass the typeRef and use `eggShell.typeName` to give a better name to the user?

Reply:
What if typeRef is not even a number? Should we do any kind of validation on that an later call `typeName`?

Reply:
Hmm since we control the creation of typeRefs internally I don't think it will be common for the typeRef to be invalid, but it is likely types will be added that are not supported by the visitor yet. How does that sound?

Maybe a sanity `typeRef === 0 ? 'invalid type' : 'eggShell.typeName(typeRef)'` is sufficient?